.github/workflows/devshell.yml

```diff
@@ -1,16 +1,77 @@
 name: CI @ devshell
 
 on:
-  pull_request:
+  pull_request_target:
   push:
   schedule:
     - cron: '00 21 * * *'
 
 jobs:
+  cache-image-devshell:
+    runs-on: ubuntu-latest
+
+    outputs:
+      docker-image: ${{ steps.determine-actual-image.outputs.docker-image }}
+
+    env:
+      CACHED_IMAGE_NAME: ghcr.io/syslog-ng/dbld-devshell:latest
+
+    steps:
+      - name: Checkout syslog-ng source
+        uses: actions/checkout@v2
+
+      - name: Determine cached image ID
+        run: |
+          ./dbld/rules pull-image-devshell || true # remove the || true
+          CACHED_IMAGE_ID=$(docker images -q ${CACHED_IMAGE_NAME})
+
+          . .github/workflows/gh-tools.sh
+          gh_export CACHED_IMAGE_ID
+
+      - name: cache-image-devshell
+        run: |
+          ./dbld/rules cache-image-devshell
+
+      - name: Determine actual image
+        id: determine-actual-image
+        run: |
+          CURRENT_IMAGE_ID=$(docker images -q ${CACHED_IMAGE_NAME})
+          if [[ "${CURRENT_IMAGE_ID}" == "${CACHED_IMAGE_ID}" ]]
+          then
+            DOCKER_IMAGE=${CACHED_IMAGE_NAME}
+            USE_CACHED_IMAGE=true
+          else
+            DOCKER_IMAGE=ghcr.io/${{ github.actor }}/dbld-devshell:ci-${{ github.run_id }}
+            USE_CACHED_IMAGE=false
+            docker tag ${CACHED_IMAGE_NAME} ${DOCKER_IMAGE}
+          fi
+
+          echo "Using image: ${DOCKER_IMAGE}"
+
+          . .github/workflows/gh-tools.sh
+          gh_export USE_CACHED_IMAGE
+          echo "::set-output name=docker-image::${DOCKER_IMAGE}"
+
+      - name: Log in to the Container registry
+        if: env.USE_CACHED_IMAGE == 'false'
+        uses: docker/login-action@v1
+        with:
+          registry: ghcr.io
+          username: ${{ github.actor }}
+          password: ${{ secrets.GITHUB_TOKEN }}
+
+      - name: Upload the temporary docker image
+        if: env.USE_CACHED_IMAGE == 'false'
+        run: |
+          docker push ${{ steps.determine-actual-image.outputs.docker-image }}
+
```

Comment:
Could we push this logic into dbld/rules itself?

I mean:
1) we already have a pull-image invoked from cache-image, and we have a separate step here. I'd love to see a single cache-image here and do anything to produce an image (potentially form cache) within dbld/rules
2) we seem to do a lot of back and forth deciding whether we are using the pulled or the rebuilt image, which is then used to decide if we need to push the temporary image. 
  - What if dbld/rules cache-image would be able to return information that we need here, without having to do a prefetch
  - also what if cache-image would be able to push to the registry, if a specific make variable is set (and if we did a rebuild)

I hope I make sense here, my point is if we push this entire logic to dbld/rules, and add a means to communicate information in a structured form, then we might be able to use the same cache-image-* target in all similar cases, avoiding the need to copy steps between github workflows files.

What do you think?